samples/matter/common/src/diagnostic/diagnostic_logs_provider.cpp

```diff
@@ -0,0 +1,184 @@
+						     Optional<uint64_t> &outTimeStamp,
+						     Optional<uint64_t> &outTimeSinceBoot)
+{
+	VerifyOrReturnError(IsValidIntent(intent), CHIP_ERROR_INVALID_ARGUMENT);
+
+	uint16_t freeSlot = mIntentMap.GetFirstFreeSlot();
+	VerifyOrReturnError(freeSlot != kInvalidLogSessionHandle, CHIP_ERROR_NO_MEMORY);
+
+	mIntentMap.Insert(freeSlot, std::move(intent));
+	outHandle = freeSlot;
+
+	ChipLogProgress(Zcl, "Starting Log collection for %s with session handle %d", getIntentStr(intent), outHandle);
+
+	return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR DiagnosticLogProvider::EndLogCollection(LogSessionHandle sessionHandle)
+{
+	VerifyOrReturnError(sessionHandle != kInvalidLogSessionHandle, CHIP_ERROR_INVALID_ARGUMENT);
+	VerifyOrReturnError(mIntentMap.Contains(static_cast<uint16_t>(sessionHandle)), CHIP_ERROR_INTERNAL);
+
+	CHIP_ERROR err = CHIP_NO_ERROR;
+
+	IntentEnum intent = mIntentMap[sessionHandle];
+
+	/* Old intent is no needed anymore*/
+	mIntentMap.Erase(static_cast<uint16_t>(sessionHandle));
+
+	/* Do the specific action for the removing intent */
+	switch (intent) {
+	case IntentEnum::kEndUserSupport:
+		/* TODO: add support for End User Intent */
+		break;
+	case IntentEnum::kNetworkDiag:
+		/* TODO: add support for End User Intent */
+		break;
+	case IntentEnum::kCrashLogs:
+		err = ClearCrashLogs();
+		VerifyOrReturnError(CHIP_NO_ERROR == err, err);
+		break;
+	default:
+		return CHIP_ERROR_INVALID_ARGUMENT;
+	}
+
+	ChipLogProgress(Zcl, "Ending Log collection for %s with session handle %d",
+			getIntentStr(mIntentMap[sessionHandle]), sessionHandle);
+
+	return err;
+}
+
+CHIP_ERROR DiagnosticLogProvider::CollectLog(LogSessionHandle sessionHandle, MutableByteSpan &outBuffer,
+					     bool &outIsEndOfLog)
+{
+	VerifyOrReturnError(sessionHandle != kInvalidLogSessionHandle, CHIP_ERROR_INVALID_ARGUMENT);
+	VerifyOrReturnError(mIntentMap.Contains(static_cast<uint16_t>(sessionHandle)), CHIP_ERROR_INTERNAL);
+
+	CHIP_ERROR err = CHIP_NO_ERROR;
+
+	switch (mIntentMap[sessionHandle]) {
+	case IntentEnum::kEndUserSupport:
+		/* TODO: add support for End User Intent */
+		break;
+	case IntentEnum::kNetworkDiag:
+		/* TODO: add support for End User Intent */
+		break;
+	case IntentEnum::kCrashLogs:
+		err = GetCrashLogs(outBuffer, outIsEndOfLog);
+		VerifyOrReturnError(CHIP_NO_ERROR == err, err, outBuffer.reduce_size(0));
+		break;
+	default:
+		return CHIP_ERROR_INVALID_ARGUMENT;
+	}
+
+	return CHIP_NO_ERROR;
+}
+
+size_t DiagnosticLogProvider::GetSizeForIntent(IntentEnum intent)
+{
+	return 0;
+}
+
+CHIP_ERROR DiagnosticLogProvider::GetLogForIntent(IntentEnum intent, MutableByteSpan &outBuffer,
+						  Optional<uint64_t> &outTimeStamp,
+						  Optional<uint64_t> &outTimeSinceBoot)
+{
+	LogSessionHandle sessionHandle = kInvalidLogSessionHandle;
+
+	CHIP_ERROR err = StartLogCollection(intent, sessionHandle, outTimeStamp, outTimeSinceBoot);
+	VerifyOrReturnError(CHIP_NO_ERROR == err, err, outBuffer.reduce_size(0));
```

Comment:
You duplicate the same error handling in the multiple places. Wouldn't be better to use VerifyOrExit and reduce size in the single place after exit label?